internal/util/secrets.go

```diff
@@ -15,78 +13,78 @@ package util
-// values from [0,X), we need to force this to be [40,126]
-var passwordCharSelector = big.NewInt(passwordCharUpper - passwordCharLower)
+// accumulate gathers n bytes from f and returns them as a string. It returns
+// an empty string when f returns an error.
+func accumulate(n int, f func() (byte, error)) (string, error) {
+	result := make([]byte, n)
 
-// GeneratePassword generates a password of a given length out of the acceptable
-// ASCII characters suitable for a password
-func GeneratePassword(length int) (string, error) {
-	password := make([]byte, length)
-	i := 0
-
-	for i < length {
-		val, err := rand.Int(rand.Reader, passwordCharSelector)
-		// if there is an error generating the random integer, return
-		if err != nil {
+	for i := range result {
+		if b, err := f(); err == nil {
+			result[i] = b
+		} else {
 			return "", err
 		}
+	}
 
-		char := byte(passwordCharLower + val.Int64())
+	return string(result), nil
+}
 
-		// if the character is in the exclusion list, continue
-		if idx := strings.IndexAny(string(char), passwordCharExclude); idx > -1 {
-			continue
-		}
+// randomCharacter builds a function that returns random bytes from class.
+func randomCharacter(random io.Reader, class string) func() (byte, error) {
+	if random == nil {
+		panic("requires a random source")
+	}
+	if len(class) == 0 {
+		panic("class cannot be empty")
+	}
+
+	size := big.NewInt(int64(len(class)))
 
-		password[i] = char
-		i++
+	return func() (byte, error) {
+		i, err := rand.Int(random, size)
+		if err != nil {
+			i = new(big.Int)
+		}
+		return class[int(i.Int64())], err
 	}
+}
+
+var randomAlphaNumeric = randomCharacter(rand.Reader, ``+
+	`ABCDEFGHIJKLMNOPQRSTUVWXYZ`+
+	`abcdefghijklmnopqrstuvwxyz`+
+	`0123456789`)
 
-	return string(password), nil
+// GenerateAlphaNumericPassword returns a random alphanumeric string.
+func GenerateAlphaNumericPassword(length int) (string, error) {
+	return accumulate(length, randomAlphaNumeric)
 }
 
-// GeneratedPasswordLength returns the value for what the length of a
-// randomly generated password should be. It first determines if the user
-// provided this value via a configuration file, and if not and/or the value is
-// invalid, uses the default value
-func GeneratedPasswordLength(configuredPasswordLength string) int {
-	// set the generated password length for random password generation
-	// note that "configuredPasswordLength" may be an empty string, and as such
-	// the below line could fail. That's ok though! as we have a default set up
-	generatedPasswordLength, err := strconv.Atoi(configuredPasswordLength)
-	// if there is an error...set it to a default
-	if err != nil {
-		generatedPasswordLength = DefaultGeneratedPasswordLength
-	}
+// policyASCII is the list of acceptable characters from which to generate an
+// ASCII password.
+const policyASCII = `` +
+	`()*+,-./` + `:;<=>?@` + `[]^_` + `{|}` +
+	`ABCDEFGHIJKLMNOPQRSTUVWXYZ` +
```

Comment:
Do you think it's worthwhile to define AlphaNumeric as a constant, then `+` it here and use in in the var randomAlphanumeric? Like, what does that save? Typing it out twice? It doesn't seem like something we need to worry about typing out multiple times in the future (esp since we can test with a regex). So, probably not worth it, but just wanted to mention it.

Reply:
In my mind, these two lists/sets are separate and unrelated. Sharing a constant would indicate to me that we intentionally want this one to be a superset of the other.